paperqa/agents/tools.py

```diff
@@ -373,6 +370,10 @@ async def complete(
         # TODO: eliminate race condition here if agent calls 2+ times in parallel
         # with opposite has_successful_answer values
         state.session.has_successful_answer = has_successful_answer
+
+        if not state.session.answer:
```

Comment:
This technically won't be needed in a normal rollout with the extra logic above -- BUT if you're calling steps from some other orchestrator, then it covers your bases.